tests/unit/test_client.py

```diff
@@ -1106,3 +1108,72 @@ def test_verify_takes_precedence_over_scoped_config(self):
                               scoped_config={'ca_bundle': 'mycerts.pem'})
         call_kwargs = self.endpoint_creator.create_endpoint.call_args[1]
         self.assertEqual(call_kwargs['verify'], 'verify.pem')
+
+
+class TestConfig(unittest.TestCase):
+    def test_can_use_args_to_construct(self):
+        config = client.Config(*client.Config.OPTION_DEFAULTS.values())
+        for option, default_value in client.Config.OPTION_DEFAULTS.items():
+            self.assertTrue(hasattr(config, option))
+            self.assertEqual(getattr(config, option), default_value)
+
+    def test_can_use_kwargs_to_construct(self):
+        config = client.Config(**client.Config.OPTION_DEFAULTS)
+        for option, default_value in client.Config.OPTION_DEFAULTS.items():
+            self.assertTrue(hasattr(config, option))
+            self.assertEqual(getattr(config, option), default_value)
+
+    def test_can_use_mix_of_args_and_kwargs(self):
+        config = client.Config('us-east-1', read_timeout=50)
+        self.assertEqual(config.region_name, 'us-east-1')
+        self.assertEqual(config.read_timeout, 50)
+
+    def test_invalid_kwargs(self):
+        with self.assertRaisesRegexp(TypeError, 'Got unexpected keyword'):
+            client.Config(foo='foo')
+
+    def test_pass_invalid_length_of_args(self):
+        with self.assertRaisesRegexp(TypeError, 'Takes at most'):
+            client.Config('foo', *client.Config.OPTION_DEFAULTS.values())
+
+    def test_create_with_multiple_kwargs(self):
+        with self.assertRaisesRegexp(TypeError, 'Got multiple values'):
+            client.Config('us-east-1', region_name='us-east-1')
+
+    def test_merge_returns_new_config_object(self):
+        config = client.Config()
+        other_config = client.Config()
+        new_config = config.merge(other_config)
+        # Check the type is correct
+        self.assertIsInstance(new_config, client.Config)
+        # Make sure the new config is a brand new config object
+        self.assertIsNot(new_config, config)
+        self.assertIsNot(new_config, other_config)
+
+    def test_general_merge_keeps_default_values(self):
+        config = client.Config()
+        other_config = client.Config()
+        config_properties = vars(config)
+        new_config = config.merge(other_config)
+        # Ensure that the values all stayed the same in the new config
+        self.assertEqual(config_properties, vars(new_config))
+
+    def test_merge_overrides_values(self):
+        config = client.Config(region_name='us-east-1')
+        other_config = client.Config(region_name='us-west-2')
+        new_config = config.merge(other_config)
+        self.assertEqual(new_config.region_name, 'us-west-2')
+
+    def test_merge_overrides_values_even_when_using_default(self):
```

Comment:
Can we also add this same test but for default integer values?  Would be nice to have a test that catches that we're not relying on `None` or a `False` value.  e.g original_timeout:30, new_timout:60, merged_value: 60.